### DISCUSSION TITLE
Track `model_logp` in other samplers

### DISCUSSION THREAD

**Discussion by eigenfoo** [pymc-devs/pymc3#5395] [Ideas]
Following up from #3134. The previous PR only added `model_logp` for NUTS and HMC. Apart from them, no other sampler computes the logp (e.g. the Metropolis samplers only computes the `delta_logp`), so there would be overhead to require these samplers to track the logp.

Quoting @junpenglao:
> I have an idea to do this also for Metropolis where we are computing the delta_logp instead of the logp. We can compute the model logp at sample 0, then from sample 1 onward we do state.logp = logp_tm1 - delta_logp which should recover the logp at time t while also cheap to compute.

We should decide what samplers should support exporting the logp (i.e. which samplers already do this, or something close to it).

**Comment by eigenfoo** [comment]
I suggest that we add `logp` to the following samplers:

- SMC: AFAICT we [compute the `logp` here](https://github.com/pymc-devs/pymc3/blob/master/pymc3/step_methods/smc.py#L159), and again, just need to percolate that up to the trace.
- Elliptical Slice: it looks like we [already compute `logp`](https://github.com/pymc-devs/pymc3/blob/master/pymc3/step_methods/elliptical_slice.py#L86), it's just a matter of percolating that information up to the trace, similar to what I did in #3121 
- Slicer: similar to the elliptical slice above.
- Metropolis: the Metropolis samplers compute `delta_logp` for each sample, so all that is required is to compute the `logp` for the first step, and then cumulatively sum the `delta_logp`s.

I'm not sure if we could also add the `logp` to the Gibbs sampler and SGMCMC, or if that would be worthwhile.

cc @junpenglao

**Comment by junpenglao** [comment]
re Gibbs:
https://github.com/pymc-devs/pymc3/blob/26267ca9ca57c2da4da74b5a230e4913dd6735e1/pymc3/step_methods/metropolis.py#L348

https://github.com/pymc-devs/pymc3/blob/26267ca9ca57c2da4da74b5a230e4913dd6735e1/pymc3/step_methods/metropolis.py#L435

**Comment by eigenfoo** [comment]
I'm confused, what is the difference between `gibbs.py` and `metropolis.py`?

**Comment by junpenglao** [comment]
the ElementwiseCategorical step in `gibbs.py` was supposedly deprecated but there is a bug in it or something... I dont remember exactly...

**Comment by junpenglao** [comment]
OK found the issue: https://github.com/pymc-devs/pymc3/issues/1563 and https://github.com/pymc-devs/pymc3/issues/2879

**Comment by ricardoV94** [comment]
Related to #3546
